include/zio/zio_device.h

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) 2019 Thomas Burdick <thomas.burdick@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_ZIO_DEVICE_H_
+#define ZEPHYR_INCLUDE_ZIO_DEVICE_H_
+
+#include <zephyr/types.h>
+#include <device.h>
+
+/**
+ * A ZIO device provides attributes, channels, and triggers for a device.
+ *
+ * Channels are data streams of input or output with attributes themselves which may be
+ * subscribed to.
+ */
+struct zio_channel_desc
+{
+  const char *name;
+  const u16_t kind;
+  const enum direction direction;
+  const u8_t bitwidth;
+  const u8_t bitshift;
+  const enum container container;
+  const enum byteorder byteorder;
+  const u8_t byteshift;
+};
+
+
+/**
+ * @brief A type tagged variant for attributes
+ */
+struct zio_variant
+{
+  enum {
+        float_,
+        double_,
+        u8_,
+        u16_,
+        u32_,
+        u64_,
+        s8_,
+        s16_,
+        s32_,
+        s64_,
+  } tag;
+  union {
+    float float_val;
+    double double_val;
+    u8_t  u8_val;
+    u16_t u16_val;
+    u32_t u32_val;
+    u64_t u64_val;
+    s8_t  s8_val;
+    s16_t s16_val;
+    s32_t s32_val;
+    s64_t s64_val;
+  } value;
+}
```

Comment:
Should `char *` be considered here in some form? It seems like a valid option for attributes as a concept. And `bool`, or can `u8_val` be a proxy for that?

Reply:
Memory management should also be considered here to make sure when these are stored in a buffer, we are only consuming 1 byte for `u8_val` and not the maximum 8-byte capacity (`double`, `s64_t`, etc.) of the union as it stands today.

Reply:
So I'm not planning on using the variant type for samples stored, for that I plan on going the iio route and tracking the element size

Reply:
Yeah, the same issue is present in the attr code I proposed ... values obviously need to be limited to an appropriate number of bytes, and the struct updated to work with a matching buffer.